airflow/jobs/scheduler_job.py

```diff
@@ -913,26 +913,23 @@ def _create_dag_runs(self, dag_models: Collection[DagModel], session: Session) -
                     creating_job_id=self.id,
                 )
                 active_runs_of_dags[dag.dag_id] += 1
-            self._update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id])
+            if self._should_update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id]):
```

Comment:
I still wonder if having `data_interval` as an optional kwarg to `_update_dag_next_dagruns` would keep this cleaner. Then this case can pass it, the others that don't can get it from `get_run_data_interval`, and we don't have to add an extra conditional and call to `calculate_dagrun_date_fields` in more than 1 place.

Reply:
e.g:
```
--- a/airflow/jobs/scheduler_job.py
+++ b/airflow/jobs/scheduler_job.py
@@ -913,11 +913,11 @@ class SchedulerJob(BaseJob):
                     creating_job_id=self.id,
                 )
                 active_runs_of_dags[dag.dag_id] += 1
-            self._update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id])
+            self._update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id], data_interval)
         # TODO[HA]: Should we do a session.flush() so we don't have to keep lots of state/object in
         # memory for larger dags? or expunge_all()
 
-    def _update_dag_next_dagruns(self, dag, dag_model: DagModel, total_active_runs) -> None:
+    def _update_dag_next_dagruns(self, dag, dag_model: DagModel, total_active_runs, data_interval = None) -> None:
         """
         Update the next_dagrun, next_dagrun_data_interval_start/end
         and next_dagrun_create_after for this dag.
@@ -931,7 +931,8 @@ class SchedulerJob(BaseJob):
             )
             dag_model.next_dagrun_create_after = None
         else:
-            data_interval = dag.get_next_data_interval(dag_model)
+            if not data_interval:
+                data_interval = dag.get_next_data_interval(dag_model)
             dag_model.calculate_dagrun_date_fields(dag, data_interval)
```

Probably safer with a sentinel instead of None, but I think it'd be okay.

Reply:
(I don't have a strong preference here though)

Reply:
No strong preference either but I might merge it tomorrow to cut rc2 and we can refactor if need be after it

Reply:
This will still have the missing dagruns issue because it's calling `dag.get_next_data_interval` when a run has finished. This will skip it. Using the latest dagrun when the run has just finished and calling `dag.get_run_data_interval` resolves the issue. The `get_next_data_interval` gives us the `next` dagrun's data interval but `get_run_data_interval` when supplied with a recent dagrun gives a dagruns data interval that sets the DagRunInfo correctly especially the `next_dagrun_create_after` for the dag. That way, the nullified `dag_model.next_dagrun_create_after` is set again properly.